src/bcf-toolkit/Converter/Bcf30/SchemaConverterToBcf21.cs

```diff
@@ -1,11 +1,296 @@
+    Model.Bcf30.DocumentReference from) {
+    var builder = new DocumentReferenceBuilder();
+    var isExternal = string.IsNullOrEmpty(from.Url);
+    builder
+      .SetGuid(from.Guid)
+      .SetDescription(from.Description)
+      .SetIsExternal(isExternal)
+      .SetReferencedDocument(isExternal ? from.Url : from.DocumentGuid);
+
+    return builder.Build();
+  }
+
+  private static Model.Bcf21.Comment ConvertComment(Model.Bcf30.Comment from) {
+    var builder = new CommentBuilder();
+
+    builder
+      .SetGuid(from.Guid)
+      .SetDate(from.Date)
+      .SetAuthor(from.Author)
+      .SetCommentProperty(from.CommentProperty)
+      .SetViewPointGuid(from.Viewpoint.Guid)
+      .SetModifiedDate(from.ModifiedDate)
+      .SetModifiedAuthor(from.ModifiedAuthor);
+
+    return builder.Build();
+  }
+
+  private static Model.Bcf21.ViewPoint ConvertViewPoint(
+    Model.Bcf30.ViewPoint from) {
+
+    var viewPoint = new ViewPoint();
+
+    viewPoint.VisualizationInfo =
+      ConvertVisualizationInfo(from.VisualizationInfo);
+
+    return viewPoint;
+  }
+
+  private static Model.Bcf21.VisualizationInfo? ConvertVisualizationInfo(
+    Model.Bcf30.VisualizationInfo? from) {
+    if (from is null)
+      return null;
+
+    var builder = new VisualizationInfoBuilder();
+
+    builder
+      .SetGuid(from.Guid)
+      .AddSelections(
+        from.Components.Selection.Select(ConvertComponent).ToList())
+      .SetVisibility(ConvertVisibility(from.Components.Visibility))
+      .AddBitmaps(from.Bitmaps.Select(ConvertBitmap).ToList())
+      .AddColorings(from.Components.Coloring.Select(ConvertColoring).ToList())
+      .AddLines(from.Lines.Select(ConvertLine).ToList())
+      .AddClippingPlanes(from.ClippingPlanes.Select(ConvertClippingPlane)
+        .ToList())
+      .SetOrthogonalCamera(ConvertOrthogonalCamera(from.OrthogonalCamera))
+      .SetPerspectiveCamera(ConvertPerspectiveCamera(from.PerspectiveCamera));
+    //TODO: SetViewSetupHints data loss
+
+    return builder.Build();
+
+  }
+
+  private static Model.Bcf21.OrthogonalCamera ConvertOrthogonalCamera(
+    Model.Bcf30.OrthogonalCamera from) {
+    var camera = new OrthogonalCamera();
+
+    camera.CameraUpVector.X = from.CameraUpVector.X;
+    camera.CameraUpVector.Y = from.CameraUpVector.Y;
+    camera.CameraUpVector.Z = from.CameraUpVector.Z;
+
+    camera.CameraDirection.X = from.CameraDirection.X;
+    camera.CameraDirection.Y = from.CameraDirection.Y;
+    camera.CameraDirection.Z = from.CameraDirection.Z;
+
+    camera.CameraViewPoint.X = from.CameraViewPoint.X;
+    camera.CameraViewPoint.Y = from.CameraViewPoint.Y;
+    camera.CameraViewPoint.Z = from.CameraViewPoint.Z;
+
+    camera.ViewToWorldScale = from.ViewToWorldScale;
+    //TODO: AspectRatio data loss
+    return camera;
+  }
+
+  private static Model.Bcf21.PerspectiveCamera ConvertPerspectiveCamera(
+    Model.Bcf30.PerspectiveCamera from) {
+    var camera = new PerspectiveCamera();
```

Comment:
Perspective camera builder should be used, or do the validation manually.

Reply:
Done.